articles/node_lifecycle.md

```diff
@@ -0,0 +1,176 @@
+
+Any service requests to a node in the inactive state will not be answered (to the caller, they will fail immediately).
+
+Entering the `inactive` state from the `created` state requires running the node's `onInitialised` method. The node uses this to set up any resources it must hold throughout its life (irrespective of if it is active or inactive). As examples, such resources may include topic publications and subscriptions, memory that is held continuously, and initialising configuration parameters.
+
+Entering the `inactive` state from the `active` state requires running the node's `onDeactivated` method. See the `active` state description for more information.
+
+Entering the `inactive` state from the `fatal error` state requires running the node's `onReset` method. See the `fatal error` state description for more information.
+
+If an error occurs while in this state that cannot be handled by the node, then the transition to the `fatal error` state is taken.
+
+The node may also exit this state by being destroyed, in which case the `destroy` transition is taken.
+
+Implementation note: Causing the transition to the destroyed state can often be mapped to deletion of the node instance.
+
+### Active state
+
+This is the main state of the node's life cycle. While in this state, the node performs any processing, responds to service requests, reads and processes data, produces output, etc.
+
+Entering the `active` state requires running the node's `onActivated` method. This is used to prepare the node for execution. This may include acquiring resources that are only held while the node is actually active, such as access to hardware. Ideally, no preparation that requires significant time (such as lengthy hardware initialisation) should be performed in this callback.
+
+To shut down a node, it must first be deactivated. This will cause the `onDeactivate` method to be called, allowing the node to clean up any resources it only holds while active.
+
+If an error that cannot be handled by the node/system occurs in this state, the transition to the `fatal error` state is taken.
+
+### Fatal Error state
+
+Known errors that can occur in a node should be considered designed behaviour of the system and so should be handled by the node where the error occurred and, if necessary, the surrounding nodes. The reason for this is that the system knows best how to handle errors that occur in its behaviour.
+
+Any fatal errors (those which the node/system was not expecting and does not know how to handle) will cause a transition to the `fatal error` state. In this state, because the node does not know what to do in the presence of the error, it is not provided any execution time. Arriving data is not processed, service requests are not responded to, and no output is produced.
+
+On entering the `fatal error` state, the node's `onError` method must be executed. This provides the node with a chance to clean up resources it was using.
```

Comment:
> It is actually not accurate. The states are important because they are a form of reporting so I definitely do not advocate for removing them. The current proposal only has an error state for fatal errors. I would prefer to bring the state machine closer to RTT on these states by having a separate Exception and FatalError states (I use the RTT names for simplicity) where:

Ok then we're in agreement about the need for an exception state which is not active. And I had suggested that there is no need for a fatal error state, but I'm not passionate about that. It just seemed to me that it was simple to the point of unnecessary since it can only result in the transition to the destroyed state.

> In addition RTT has a RuntimeError state that means "I am active but in an error state". This is the state that I never saw used.

Ok, then I misunderstood. That state, as you've described it, doesn't make sense to me either.

> Why is that ? The point of having a node is to standardize the behaviour of a non-standard object (the device) into something that is standard and understood by all. Node design best practices are always driven by the desired system design.

I agree, we are doing this to get convention in place which allows us to automate and lets people write more similar code which improves reuse. I was just trying to say that if the conventions result in forcing people to write their code in a particular way, then the reasoning for that convention should be justified (not arbitrary). At the time I thought you were talking about not having an exception state, which would force people to do all error resolution in the active state or via a reset of the entire node, which I thought was too much. Since we both agree there should be an error state which is separate from the active state and recoverable, there isn't an issue.